File: krass_jass/rollout.py
"""The rollout kernel — the seam.

**This module is the one that gets replaced.** ``docs/plan-review.md`` §1 shows that M5
needs a teacher roughly 100× the serve budget, which pure Python cannot deliver; the
escalation path is Numba/Cython here, then a Rust core via PyO3. That swap is only cheap if
this file stays narrow, so the rules are:

* ints and tuples only — no dataclasses, no enums, no allocation in the loop;
* everything it needs arrives in a ``kernel`` tuple built once, outside the loop;
* it depends on :mod:`krass_jass.legal` and nothing else in the package.

It deliberately does *not* re-implement legal-move generation. One rule implementation,
one place to be wrong.
"""

from __future__ import annotations

from random import Random

from .cards import NUM_RANKS, card_suit
from .legal import legal_moves
from .rules import Contract, RulesConfig
from .scoring import TRICKS_PER_ROUND
from .tables import CARD_VALUES, STRENGTH
from .trick import NUM_SEATS

#: ``(trump, values, strength, strict_undertrump, puur_exempt, last_trick_bonus, match_bonus)``
Kernel = tuple


def make_kernel(contract: Contract, cfg: RulesConfig) -> Kernel:
    """Flatten a contract and config into the plain tuple the kernel reads.

    Build once per search, not once per rollout.
    """
    return (
        contract.trump_suit,
        CARD_VALUES[contract],
        STRENGTH[contract],
        cfg.strict_undertrump,
        cfg.puur_exempt_trump_lead,
        cfg.last_trick_bonus,
        cfg.match_bonus,
    )


def play_out(hands: list[int], leader: int, kernel: Kernel, rng: Random) -> tuple[int, int]:
    """Play out at random from ``leader`` until the hands are empty. Returns team points.

    Works from a partial position as well as from the deal — which is the whole point,
    since DMCTS rolls out from wherever the search currently is. **Precondition:** the
    position is at a trick boundary, so every hand holds the same number of cards. (Rolling
    out from mid-trick needs the cards already on the table; that arrives with the search
    in M4.)

    ``hands`` is consumed (the caller passes a copy). Points include the last-trick bonus,
    and the match bonus only when the playout covered a whole round — from a partial
    position the kernel cannot know who took the earlier tricks, so the caller owns it.
    Multiplier, Weis and Stöck are applied outside; here they would only scale the
    comparison.
    """
    trump, values, strength_by_led, strict, puur, last_bonus, match_bonus = kernel
    pts = [0, 0]
    tricks = [0, 0]
    randrange = rng.randrange
    tricks_to_play = hands[leader].bit_count()

    for _ in range(tricks_to_play):
        led = -1
        best_trump = -1
        strength = None
        trick_pts = 0
        best_seat = leader
        best_str = -1

        for i in range(NUM_SEATS):
            seat = (leader + i) & 3
            hand = hands[seat]
            legal = legal_moves(
                hand, trump, led, best_trump, strict_undertrump=strict, puur_exempt=puur
            )

            # Uniform pick over set bits, without materialising a list.
            n = legal.bit_count()
            m = legal
            for _ in range(randrange(n)):
                m &= m - 1
            low = m & -m
            c = low.bit_length() - 1

            hands[seat] = hand ^ low
            trick_pts += values[c]

            if i == 0:
                led = card_suit(c)
                strength = strength_by_led[led]
                best_str = strength[c]
            else:
                s = strength[c]
                if s > best_str:
                    best_str, best_seat = s, seat
            if trump >= 0 and c // NUM_RANKS == trump:
                ts = strength[c] - 100
                if ts > best_trump:
                    best_trump = ts

        team = best_seat & 1
        pts[team] += trick_pts
        tricks[team] += 1
        leader = best_seat

    if not tricks_to_play:
        return 0, 0
    pts[leader & 1] += last_bonus
    if match_bonus and tricks_to_play == TRICKS_PER_ROUND:
        for t in (0, 1):
            if tricks[t] == TRICKS_PER_ROUND:
                pts[t] += match_bonus
    return pts[0], pts[1]

File: rust/src/rollout.rs
//! The rollout kernel and trick resolution.
//!
//! This is the code the whole Rust port exists for. `docs/plan-review.md` §1 measured the
//! Python version at ~72k rollouts/sec; the search on top of it ran at 35k iterations/sec,
//! with 73% of time in here. Note that porting *only* this would have been worth ~2.4x —
//! the tree in `search.rs` had to come with it.

use crate::cards::*;
use crate::legal::legal_moves;
use crate::tables::{CARD_VALUES, STRENGTH};
use crate::rng::Rng;

/// Everything the kernel needs, flattened once outside the loop.
#[derive(Clone, Copy)]
pub struct Kernel {
    pub contract: usize,
    pub trump: i32,
    pub strict_undertrump: bool,
    pub puur_exempt: bool,
    pub last_trick_bonus: i32,
    pub match_bonus: i32,
}

impl Kernel {
    pub fn new(
        contract: usize,
        strict_undertrump: bool,
        puur_exempt: bool,
        last_trick_bonus: i32,
        match_bonus: i32,
    ) -> Self {
        Kernel {
            contract,
            trump: if contract < 4 { contract as i32 } else { -1 },
            strict_undertrump,
            puur_exempt,
            last_trick_bonus,
            match_bonus,
        }
    }
}

/// Uniform pick over the set bits of `mask`. Returns the card index.
#[inline(always)]
pub fn pick_random(mask: u64, rng: &mut Rng) -> usize {
    let n = mask.count_ones();
    let mut m = mask;
    let skip = rng.below(n);
    for _ in 0..skip {
        m &= m - 1;
    }
    m.trailing_zeros() as usize
}

/// Play out at random until the hands are empty. Returns team points.
///
/// Precondition, as in the Python kernel: the position is at a trick boundary, so every
/// hand holds the same number of cards. The match bonus is only awarded when the playout
/// covered a whole round — from a partial position the kernel cannot know who took the
/// earlier tricks.
pub fn play_out(hands: &mut [u64; NUM_SEATS], mut leader: usize, k: &Kernel, rng: &mut Rng) -> (i32, i32) {
    let values = &CARD_VALUES[k.contract];
    let strength_by_led = &STRENGTH[k.contract];

    let n_tricks = hands[leader].count_ones() as usize;
    if n_tricks == 0 {
        return (0, 0);
    }

    let mut pts = [0i32; 2];
    let mut tricks = [0usize; 2];

    for _ in 0..n_tricks {
        let mut led: i32 = -1;
        let mut best_trump: i32 = -1;
        let mut strength: &[i32; NUM_CARDS] = &strength_by_led[0];
        let mut trick_pts = 0i32;
        let mut best_seat = leader;
        let mut best_str = -1i32;

        for i in 0..NUM_SEATS {
            let seat = (leader + i) & 3;
            let hand = hands[seat];
            let legal = legal_moves(
                hand,
                k.trump,
                led,
                best_trump,
                k.strict_undertrump,
                k.puur_exempt,
            );
            let c = pick_random(legal, rng);

            hands[seat] = hand ^ (1u64 << c);
            trick_pts += values[c];

            if i == 0 {
                led = card_suit(c) as i32;
                strength = &strength_by_led[led as usize];
                best_str = strength[c];
            } else {
                let s = strength[c];
                if s > best_str {
                    best_str = s;
                    best_seat = seat;
                }
            }
            if k.trump >= 0 && card_suit(c) as i32 == k.trump {
                let ts = strength[c] - 100;
                if ts > best_trump {
                    best_trump = ts;
                }
            }
        }

        let team = best_seat & 1;
        pts[team] += trick_pts;
        tricks[team] += 1;
        leader = best_seat;
    }

    pts[leader & 1] += k.last_trick_bonus;
    if k.match_bonus != 0 && n_tricks == TRICKS_PER_ROUND {
        for t in 0..2 {
            if tricks[t] == TRICKS_PER_ROUND {
                pts[t] += k.match_bonus;
            }
        }
    }
    (pts[0], pts[1])
}
